src/common.rs

```diff
@@ -367,6 +369,54 @@ pub enum RelocationEncoding {
     ///
     /// The `RelocationKind` must be PC relative.
     LoongArchBranch,
+    
+    /// SHARC+ 24-bit absolute address in a Type A instruction
+    /// 
+    /// Used with `RelocationKind::Absolute`.
+    /// See [elf::R_SHARC_ADDR24_v3]
+    SharcAbs24TypeA,
+    
+    /// SHARC+ 32-bit absolute address in a Type A instruction
+    /// 
+    /// Used with `RelocationKind::Absolute`.
+    /// See [elf::R_SHARC_ADDR32_v3]
+    SharcAbs32TypeA,
+    
+    /// SHARC+ 32-bit absolute address in a 32-bit STT_OBJECT location
+    /// 
+    /// See [elf::R_SHARC_ADDRVAR_V3]
+    SharcAbs32Data,
+    
+    /// SHARC+ 6-bit relative address in a Type A instruction
+    /// 
+    /// The `RelocationKind` must be PC relative.
+    SharcPcr6TypeA,
+    
+    /// SHARC+ 24-bit relative address in a Type A instruction
+    /// 
+    /// The `RelocationKind` must be PC relative.
+    SharcPcr24TypeA,
+    
+    /// SHARC+ 6-bit absolute address in the immediate value field of a Type A instruction
+    SharcAbs6TypeA,
+    
+    /// SHARC+ 16-bit absolute address in the immediate value field of a Type A instruction
+    SharcAbs16TypeA,
+    
+    /// SHARC+ 6-bit absolute address in the immediate value field of a Type B instruction
+    SharcAbs6TypeB,
+    
+    /// SHARC+ 7-bit absolute address in the immediate value field of a Type B instruction
+    SharcAbs7TypeB,
+    
+    /// SHARC+ 16-bit absolute address in a Type B instruction
+    SharcAbs16TypeB,
+    
+    /// SHARC+ 6-bit relative address in a Type B instruction
+    SharcPcr6TypeB,
+    
+    /// SHARC+ 16-bit absolute address in a 16-bit STT_OBJECT location
+    SharcAbs16Data,
```

Comment:
This crate does relocations a bit different from other software, and uses a combination of 3 fields to describe a relocation: kind, encoding, and size. It seems like these encodings are duplicating information for kind (absolute, PC-relative), and size (6, 7, 16, 24). Is it possible to reduce this to a smaller number of encodings? e.g TypeA and TypeB. Also I think that SharcAbs32Data and SharcAbs16Data probably should use Generic instead?

Reply:
I was confused about the details of `size`, and from your comment it looks like I guessed wrong. I've used the total bit length of the relocated instruction, but from your comment it looks like it means the length of the field within the instruction which contains the relocated address?

Reply:
Updated. Size is fixed, and I now only add `SharcTypeA` & `SharcTypeB` variants. Also ran rustfmt against previous commits, which is why they're changed.

Reply:
The documentation for `Relocation::offset` and `Relocation::size` is a bit unclear currently, but I think they necessarily depend on the encoding used, and you get to choose what they mean when adding the encoding. The implementation for AArch64 and LoongArch branches has used the offset of the instruction and the bit length of the field within the instruction. Everything else is currently a whole number of bytes, and uses the offset and length of the field, even if the field is part of larger instruction.

If you agree that using `SharcTypeA` and `SharcTypeB` makes sense, then I think doing that would be good for consistency with the other architectures. On the other hand, if this doesn't make much sense, then it may be better to use `RelocationKind::Elf` instead (and ignore the encoding and size values; once #585 is merged you would use `RelocationFlags::Elf` and wouldn't specify kind/encoding/size at all). The purpose of kind/encoding/size is to allow you to write code that is independent of file format and architecture, and mostly such code only deals with relocations that have `RelocationEncoding::Generic`, so we wouldn't be losing much by not using kind/encoding/size for other SHARC relocations.